### PR TITLE
docs(federation): describe the technical federation architecture

### PR DESCRIPTION
**Pull request by clauspeterhuebner**
### Issues

- None

### Todo

- [X] draw a picture of the technical architecture for the current gradiod system moduls
- [X] describe the steps and communication between the gradido system moduls of the current federation process

### CODE REVIEW

**Review by ogerly — DISMISSED** [review]

**Review by Mogge — APPROVED** [review]

**Comment by Mogge** [comment]
@einhornimmond Thank you for the improvement of the graphic!

File: docu/Concepts/TechnicalRequirements/Federation.md [hunk 1]
```diff
@@ -41,12 +18,28 @@ To enable such a relationship between an existing community and a new community
...
-### Overview
+### Overview of Federation-Handshake
 
 At first the following diagramm gives an overview of the three stages and shows the handshake between an existing community-A and a new created community-B including the data exchange for buildup such a federated, authenticated and autorized relationship.
 
 ![FederationHyperSwarm.png](./image/FederationHyperSwarm.png)
 
+### Technical Architecture
+
+The previous described handshake will be done by several technical moduls of the gradido system. The following picture gives an overview about the moduls and how the communicate with each other.
```
Comment (ulfgebhardt) [inline docu/Concepts/TechnicalRequirements/Federation.md:29]:
```suggestion
The previous described handshake will be done by several technical modules of the gradido system. The following picture gives an overview about the modules and how the communicate with each other.
```

File: docu/Concepts/TechnicalRequirements/Federation.md [hunk 2]
```diff
@@ -41,12 +18,28 @@ To enable such a relationship between an existing community and a new community
...
+
+![img](./image/TechnicalOverview_V1-19.svg)
+
+As soon as a Gradido Community is up and running the DHT-Modul first write the home-community-entries in the database and starts with the federation per HyperSwarm. Each community, which is configured to listen on the GRADIDO_HUB of the DHT will be part of the Gradido-Net-Federation. That means each DHT-Modul of each community will receive the publicKey of all connected communities. The DHT-Modul will open for each received publicKey a communication-socket with the associated community DHT-Modul, to exchange api-version info and hosted urls for later direct communication between both communities. The exchanged api-version info and urls will be written in the own database.
```
Comment (ulfgebhardt) [inline docu/Concepts/TechnicalRequirements/Federation.md:33]:
```suggestion
As soon as a Gradido Community is up and running the DHT-Modul first writes the home-community-entries in the database and starts with the federation via HyperSwarm. Each community, which is configured to listen on the GRADIDO_HUB of the DHT will be part of the Gradido-Net-Federation. That means each DHT-Modul of each community will receive the publicKey of all connected communities. The DHT-Modul will open for each received publicKey a communication-socket with the associated community DHT-Modul, to exchange api-version info and hosted urls for later direct communication between both communities. The exchanged api-version info and urls will be written in the own database.
```

`GRADIDO_HUB` is variable in this case and just the default value - more correct would be `a topic` or similar description.

Event tho I know what `hosted urls` is referring to, I believe someone without knowledge about this process will have trouble to understand.

File: docu/Concepts/TechnicalRequirements/Federation.md [hunk 3]
```diff
@@ -41,12 +18,28 @@ To enable such a relationship between an existing community and a new community
...
+
+The up and running Backend-Modul contains a validation logic to verify the community entries from the own DHT-Modul. For each announced but unverified community-entry the GraphQL-Client is used to invoke a getPublicKey-Request. Depending on the containing api-version the matching GraphQL-Client is used and the getPublicKey-Request will be send to the given URL.
+
+As soon as the FederationModul of the assoziated community received the getPublicKey-request the own publicKey is read from database and send back in the response.
```
Comment (ulfgebhardt) [inline docu/Concepts/TechnicalRequirements/Federation.md:37]:
```suggestion
As soon as the FederationModul of the associated community received the getPublicKey-request the own publicKey is read from database and send back in the response.
```

File: docu/Concepts/TechnicalRequirements/Federation.md [hunk 4]
```diff
@@ -41,12 +18,28 @@ To enable such a relationship between an existing community and a new community
...
+
+The GraphQL-Client will read from the returned response data the publicKey of the other community and compare it with the data of the community-entry, which cause the getPublicKey-Request. If they match the community-entry will be updated be inserting the current timestamp in the verifiedAt-field of this community-entry.
```
Comment (ulfgebhardt) [inline docu/Concepts/TechnicalRequirements/Federation.md:39]:
```suggestion
The GraphQL-Client will read the publicKey of the other community from the returned response data and compare it with the data of the community-entry, which caused the getPublicKey-Request. If they match the community-entry will be updated be inserting the current timestamp in the verifiedAt-field of this community-entry.
```

**Comment by ulfgebhardt** [comment]
Besides some types I believe this is ready to merge.

**Review by Elweyn — CHANGES_REQUESTED** [review]

File: docu/Concepts/TechnicalRequirements/Federation.md (hunk 2 above)
Comment (clauspeterhuebner) [reply docu/Concepts/TechnicalRequirements/Federation.md:33]: Yes that's correct. i added a more detailed description about the background of the exchanged data api-version and url.
good hint...

**Review by ogerly — APPROVED** [review]

**Review by Elweyn — APPROVED** [review]

**Pull request merged**